### PR TITLE
working on new getpot parser

### PR DESCRIPTION
**Pull request by rwcarlsen**
A more strict+robust getpot parser with much nicer error messages.

### CODE REVIEW

**Comment by dschwen** [comment]
Why are you rewriting all those input files. I noticed that one transformation is to combine split blocks. This will break some documentation pages. I have purposely split blocks to include simplified subsections of test files in some cases.

**Comment by rwcarlsen** [comment]
@dschwen - this says "do not review".  When I'm ready, I'll ping for discussion.  Since things in the parser will be changing a lot, it will be super easy to us to decide together what we want and adjust behavior accordingly.

**Comment by dschwen** [comment]
This was not a "review". I just don't want you to waste time on changes that you may not be aware of being breaking.

**Comment by rwcarlsen** [comment]
@idaholab/moose-team Merge anyone?

File: framework/contrib/hit/Makefile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+
+ifeq ($(CXX),)
+	CXX=g++
```
Comment (dschwen) [inline framework/contrib/hit/Makefile]: `CXX ?= g++`

File: framework/contrib/hit/hit.h [hunk 2]
```diff
@@ -0,0 +1,9 @@
+
+#ifndef HIT_HIT
+#define HIT_HIT
+
+#include "parse.h"
+
+#include <fstream> // TODO: remove this - hit parser doesn't need it at all
```
Comment (dschwen) [inline framework/contrib/hit/hit.h]: Why add it then?

File: framework/contrib/hit/lex.cc [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+  return tokTypeName(type) + ":'" + val + "'";
+}
+
+Lexer::Lexer(std::string name, std::string input) : _name(name), _input(input) {}
+std::vector<Token>
+Lexer::tokens()
+{
+  return _tokens;
+}
+std::vector<Token>
```
Comment (dschwen) [inline framework/contrib/hit/lex.cc]: empty line

File: framework/contrib/hit/lex.cc [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+Lexer::run(LexFunc start)
+{
+  LexFunc state = start;
+  while (state != NULL)
```
Comment (dschwen) [inline framework/contrib/hit/lex.cc]: `nullptr`?

File: framework/contrib/hit/lex.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+private:
+  std::string _name;
+  std::string _input;
+  size_t _start = 0;
+  size_t _pos = 0;
+  int _width = 0;
+  std::vector<Token> _tokens;
+};
+
+#define EOF (-1)
```
Comment (dschwen) [inline framework/contrib/hit/lex.h]: I think the actual value is _implementation dependent_. You might want to save that into a different pp symbol, or just rename your enum entry...

File: framework/contrib/hit/lex.h [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  TokType type;
+  /// val is the actual text from the input that makes this token.
+  std::string val;
+  /// offset is the byte offset into the original input identifying the start position where this
+  /// token was found.  This can be used to determine line numbers, column offsets, etc. useful for
+  /// error messages among other things.
+  size_t offset;
+  /// line is the line number of the original input (lines deliminted by the unix newline '\n'
+  /// character) on which the beginning of the token was found - this is redundant with the offset
+  /// combined with a reference to the original input, but is here for convenience.
```
Comment (dschwen) [inline framework/contrib/hit/lex.h]: we use 
```
/**
 *
 */
```

For multiline Doxygen comments

File: modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i [hunk 7]
```diff
@@ -44,13 +44,20 @@
...
-  displacements = 'disp_x disp_y'
-  add_variables = true
-  strain = SMALL
-  incremental = true
-  eigenstrain_names = eigenstrain
-  generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_yz stress_zx'
+[Modules]
+    [TensorMechanics]
+        [Master]
+            displacements = 'disp_x disp_y'
```
Comment (dschwen) [inline modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i]: why is that line here?

Comment (dschwen) [reply modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i]: Are you not supporting the previous compact syntax with `/`s ?

File: modules/solid_mechanics/test/tests/j_integral/tests [hunk 8]
```diff
@@ -58,18 +58,18 @@
  [../]
  [./j_3d_as_2d_topo_q_outq]
    type = 'Exodiff'
-   input = 'j_integral_3d_as_2d_topo_q_func.i cli_args = "DomainIntegral/output_q=false" "Outputs/file_base=j_integral_3d_as_2d_topo_q_func_outputQ_out"'
+   input = 'j_integral_3d_as_2d_topo_q_func.i "DomainIntegral/output_q=false" "Outputs/file_base=j_integral_3d_as_2d_topo_q_func_outputQ_out"'
```
Comment (dschwen) [inline modules/solid_mechanics/test/tests/j_integral/tests]: This does not look like the right fix

File: modules/tensor_mechanics/test/tests/tensile/tests [hunk 9]
```diff
@@ -12,7 +12,6 @@
     csvdiff = 'small_deform1.csv'
     rel_err = 1.0E-5
     abs_zero = 1.0E-5
-    cli_args = 'GlobalParams/volumetric_locking_correction=true'
```
Comment (dschwen) [inline modules/tensor_mechanics/test/tests/tensile/tests]: Why are you removing all those? This does not look like parser related, but an actual physics change.

File: test/tests/bcs/vectorpostprocessor/vectorpostprocessor.i [hunk 10]
```diff
@@ -1,8 +1,3 @@
-Nu = 4
-k = 1
-half_channel_length = 0.5
-h = ${/ ${* ${Nu} ${k}} ${half_channel_length}}
```
Comment (dschwen) [inline test/tests/bcs/vectorpostprocessor/vectorpostprocessor.i]: not supported anymore? :-/

File: test/tests/kernels/forcing_function/forcing_function_error_check.i [hunk 11]
```diff
@@ -17,7 +17,7 @@
 
   [./forcing_func]
     type = ParsedFunction
-    value = "alpha*alpha*pi*pi*sin(alpha*pi*x)"
+    value = '"alpha*alpha*pi*pi*sin(alpha*pi*x)"'
```
Comment (dschwen) [inline test/tests/kernels/forcing_function/forcing_function_error_check.i]: Not the right fix. Remove `"`

File: test/tests/misc/check_error/missing_req_par_mesh_block_test.i [hunk 12]
```diff
@@ -12,15 +12,13 @@
   nx = 3
   ny = 3
   elem_type = QUAD4
-
-  strips = 3  # this is a required params (miss-spelled) => expect to error out
```
Comment (dschwen) [inline test/tests/misc/check_error/missing_req_par_mesh_block_test.i]: why remove this? looks like it tests an error

File: framework/src/parser/Parser.C [hunk 13]
```diff
@@ -50,6 +50,7 @@
 #include <fstream>
 #include <iomanip>
 #include <algorithm>
+#include <stdlib.h>
```
Comment (dschwen) [inline framework/src/parser/Parser.C]: `<cstdlib>`?

File: framework/contrib/hit/hit.h (hunk 2 above)
Comment (rwcarlsen) [reply framework/contrib/hit/hit.h]: Lots of moose and app files were borrowing the fstream include from the old getpot parser's header :-(.  So removing this will be a fix for another day.

File: framework/contrib/hit/lex.h (hunk 5 above)
Comment (rwcarlsen) [reply framework/contrib/hit/lex.h]: I know - I couldn't decide on a satisfactory solution, so the code just stayed like this :-/

File: modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i (hunk 7 above)
Comment (rwcarlsen) [reply modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i]: It is supported, but there is a long story about getting bison tests to pass and dealing with quirks of documentation system using the old parser still, etc. - so this needs to be like this (until we get python parser switched/converted)

File: modules/solid_mechanics/test/tests/j_integral/tests (hunk 8 above)
Comment (rwcarlsen) [reply modules/solid_mechanics/test/tests/j_integral/tests]: What do you think it should be?  It passes.

File: modules/tensor_mechanics/test/tests/tensile/tests (hunk 9 above)
Comment (rwcarlsen) [reply modules/tensor_mechanics/test/tests/tensile/tests]: Unused parameters.

File: test/tests/bcs/vectorpostprocessor/vectorpostprocessor.i (hunk 10 above)
Comment (rwcarlsen) [reply test/tests/bcs/vectorpostprocessor/vectorpostprocessor.i]: plain substitutions are still supported, but we decided we want to be deliberate and careful about introducing and potentially redesigning nifty functionality like this.  So out for now.

File: test/tests/kernels/forcing_function/forcing_function_error_check.i (hunk 11 above)
Comment (rwcarlsen) [reply test/tests/kernels/forcing_function/forcing_function_error_check.i]: weird, not sure what I was thinking here...

File: test/tests/misc/check_error/missing_req_par_mesh_block_test.i (hunk 12 above)
Comment (rwcarlsen) [reply test/tests/misc/check_error/missing_req_par_mesh_block_test.i]: Because the sequence of error processing/handling with the new parser is a bit different and another error in this file always triggers first.   Bla, bla, bla - it was carefully and intentionally changed :-)

File: modules/solid_mechanics/test/tests/j_integral/tests (hunk 8 above)
Comment (dschwen) [reply modules/solid_mechanics/test/tests/j_integral/tests]: Maybe

```
input = j_integral_3d_as_2d_topo_q_func.i 
cli_args = 'DomainIntegral/output_q=false Outputs/file_base=j_integral_3d_as_2d_topo_q_func_outputQ_out'
```

Tacking the CLI args onto the input file name parameter seems rather ugly.

File: test/tests/misc/check_error/missing_req_par_mesh_block_test.i (hunk 12 above)
Comment (dschwen) [reply test/tests/misc/check_error/missing_req_par_mesh_block_test.i]: :-|

File: modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i (hunk 7 above)
Comment (dschwen) [reply modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i]: I bet it is long. You are not changing the old parser, so why do you need to change inputs for something that _uses_ the old parser? And why do we care about BISON tests in the combined module tests? If this is just going to be another "Bla, bla, bla" answer then don't bother, just let my dubiety be noted.

Comment (rwcarlsen) [reply modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i]: The documentation system reads this file using the old parser.  With this PR, MOOSE reads this file using the new parser.  So we needed to find a version of the input that fit in the intersection of what the old and new parsers consider correct.

Comment (sapitts) [reply modules/combined/test/tests/elastic_thermal_patch/elastic_thermal_jacobian_rz_smp.i]: We care about the BISON input files that use the tensor mechanics master action so that we can successfully pull in the relevant snippets in the BISON documentation; without somesort of work around, the documentation system will pull in the entire input file when any block below the master action is specified in the documentation system.  @dschwen I'd be happy to demonstrate this problem for you

File: modules/tensor_mechanics/test/tests/tensile/tests (hunk 9 above)
Comment (sapitts) [reply modules/tensor_mechanics/test/tests/tensile/tests]: volumetric locking correction is false by default in tensor mechanics

File: framework/contrib/hit/lex.h (hunk 6 above)
Comment (permcody) [reply framework/contrib/hit/lex.h]: I know this is contrib so it doesn't necessarily have to look the same as MOOSE, but the block style does look much nicer where you have these large text blocks.

File: unit/src/HitTests.C [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    if (n->strVal() != test.val)
+    {
+      FAIL() << "case " << i + 1 << " wrong value (key=" << test.key << "): got '" << n->strVal()
+             << "', want '" << test.val << "'\n";
+      continue;
+    }
+
+    auto f = dynamic_cast<hit::Field *>(n);
+    if (f->kind() != test.kind)
```
Comment (permcody) [inline unit/src/HitTests.C]: Looks like you routinely use the dynamic_cast, then use the resulting type directly without checking it. Dynamic cast are usually used where the cast may or may not succeed, but in your case you are definitely assuming it does. A static cast might be a better choice.

File: framework/src/parser/Parser.C [hunk 15]
```diff
@@ -69,58 +69,157 @@ Parser::Parser(MooseApp & app, ActionWarehouse & action_wh)
...
-  bool retValue = false;
-  auto found = s.find_last_of('/');
+  hit::Node * n = root->find(path);
+  while (n != nullptr)
+  {
+    hit::Node * section = n->parent();
+    if (section)
+    {
+      auto an = section->find("active");
+      auto ian = section->find("inactive");
```
Comment (permcody) [inline framework/src/parser/Parser.C]: Not a huge fan of these names. They are bad. If you start in the code below, these names mean nothing.

File: framework/src/parser/Parser.C [hunk 16]
```diff
@@ -69,58 +69,157 @@ Parser::Parser(MooseApp & app, ActionWarehouse & action_wh)
...
 
-  // Base Level is always active (skip the rest of the routine)
-  if (found == std::string::npos)
-    return true;
+      // only check current level, not nested ones
+      if (an && an->type() == hit::NodeType::Field && an->parent() == section)
+      {
+        auto vars = section->param<std::vector<std::string>>("active");
+        bool havevar = false;
```
Comment (permcody) [inline framework/src/parser/Parser.C]: `have_var` for compound variable names. This is in the style guide.

File: framework/src/parser/Parser.C [hunk 17]
```diff
@@ -69,58 +69,157 @@ Parser::Parser(MooseApp & app, ActionWarehouse & action_wh)
...
+  void walk(std::string fullpath, std::string /*nodepath*/, hit::Node * n)
+  {
+    std::string prefix = n->type() == hit::NodeType::Field ? "parameter" : "section";
+
+    if (_have.count(fullpath) > 0)
+    {
+      auto hn = _have[fullpath];
+      if (_duplicates.count(fullpath) == 0)
+      {
+        errors.push_back(errormsg(_fname, hn, prefix, " '", fullpath, "' supplied multiple times"));
```
Comment (permcody) [inline framework/src/parser/Parser.C]: `hn` is not a good variable name.

File: framework/contrib/hit/lex.cc [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+const std::string allspace = " \t\n\r";
+const std::string newline = "\n\r";
+const std::string alphanumeric = digits + alpha;
+const std::string identchars = alphanumeric + "_./:<>-+";
+
+_LexFunc::_LexFunc(LexFunc pp) : p(pp) {}
+_LexFunc::operator LexFunc() { return p; }
+
+bool
+charIn(char c, std::string valid)
```
Comment (permcody) [inline framework/contrib/hit/lex.cc]: Any reason you can't use `const std::string &` throughout? It appears that the entire parser uses string copies everywhere. I know parsing isn't a critical region, but we typically use const references in most of MOOSE. Something to consider.

File: framework/src/parser/Parser.C [hunk 19]
```diff
@@ -139,425 +238,320 @@ Parser::getFileName(bool stripLeadingPath) const
...
-  /**
-   * Set the class variable to indicate that sections names have been read, this is used later by
-   * the checkOverriddenParams function
-   */
-  _sections_read = true;
-
-  // A map containing all of the section names and active/inactive subblocks
-  std::map<std::string, BlockLists> block_lists;
-  for (auto & section_name : section_names)
+  for (std::multimap<std::string, Syntax::ActionInfo>::iterator it = iters.first;
```
Comment (permcody) [inline framework/src/parser/Parser.C]: minor: `auto`?

File: framework/src/parser/Parser.C [hunk 20]
```diff
@@ -139,425 +238,320 @@ Parser::getFileName(bool stripLeadingPath) const
...
-  mooseAssert(last + 1 >= 0, "index \"last\" is negative");
-
-  // Remove the deleted items
-  all_vars.resize(last + 1);
-
-  std::set<std::string> difference;
-
-  std::sort(all_vars.begin(), all_vars.end());
+  // Save the filename
+  char abspath[PATH_MAX + 1];
```
Comment (permcody) [inline framework/src/parser/Parser.C]: I don't see where PATH_MAX is defined, but it looks like a compile time constant. You could use `std::array` here. We generally avoid raw c-arrays in MOOSE.

File: framework/src/parser/Parser.C [hunk 21]
```diff
@@ -139,425 +238,320 @@ Parser::getFileName(bool stripLeadingPath) const
...
-      std::set<std::string>::iterator curr = j++;
-      if (curr->find(inactive_string) != std::string::npos)
-        difference.erase(curr);
-    }
-
-  std::set<std::string> requested_vars = _getpot_file.get_requested_variables();
+  std::ifstream f(_input_filename);
+  std::stringstream ss;
+  ss << f.rdbuf();
+  std::string input = ss.str();
```
Comment (permcody) [inline framework/src/parser/Parser.C]: This is a weird idiom. I know what you are doing, but it seems a little nonstandard. I googled "ifstream read in whole file" and came across this: https://stackoverflow.com/questions/2602013/read-whole-ascii-file-into-c-stdstring, it seems there are other perhaps better ways of reading in the whole buffer.

File: framework/src/parser/Parser.C [hunk 22]
```diff
@@ -139,425 +238,320 @@ Parser::getFileName(bool stripLeadingPath) const
...
+
+  if (_errmsg.size() > 0)
+    mooseError(_errmsg);
+
+  // There are a few order dependent actions that have to be built first in
+  // order for the parser and application to function properly:
+  //
+  // SetupDebugAction: This action can contain an option for monitoring the parser progress. It must
+  // be parsed first
+  //                   to capture all of the parsing output.
```
Comment (permcody) [inline framework/src/parser/Parser.C]: Flow got kind of messed up here when we changed line length. Perhaps you can fix it?

File: framework/src/parser/Parser.C [hunk 23]
```diff
@@ -139,425 +238,320 @@ Parser::getFileName(bool stripLeadingPath) const
...
+      _syntax.getSyntaxByAction("SetupDebugAction", "setup_debug"),
+      _syntax.getSyntaxByAction("GlobalParamsAction", "set_global_params"),
+      _syntax.getSyntaxByAction("DynamicObjectRegistrationAction", "dynamic_object_registration"),
+  };
+
+  // walk all the sections extracting paramters from each into InputParameters objects
+  for (auto & sec : _secs_need_first)
+  {
+    auto n = _root->find(sec);
+    if (n != nullptr)
```
Comment (permcody) [inline framework/src/parser/Parser.C]: I'm not sure about your personal preference, but you don't need to feel obligated to explicitly test against `nullptr`.

File: framework/src/parser/Parser.C [hunk 24]
```diff
@@ -139,425 +238,320 @@ Parser::getFileName(bool stripLeadingPath) const
...
 }
 
+// Checks the input and the way it has been used and emits any errors/warnings.
+// This has to be a separate function because for we don't know if some parameters were unused
+// until all the multiapps/subapps have been fully initialized - which isn't complete until
+// *after* all the other member functions on Parser have been run.  So this is here to be
+// externally called at the right time.
 void
-Parser::appendAndReorderSectionNames(std::vector<std::string> & section_names)
+Parser::errcheck(const Parallel::Communicator & comm, bool warn_unused, bool err_unused)
```
Comment (permcody) [inline framework/src/parser/Parser.C]: `errcheck` -> `errorCheck`

File: framework/src/parser/Parser.C [hunk 25]
```diff
@@ -797,76 +785,78 @@ using std::string;
...
-                                            bool in_global,
-                                            GlobalParamsAction * global_block);
-
-template <>
-void Parser::setScalarParameter<Point>(const std::string & full_name,
-                                       const std::string & short_name,
-                                       InputParameters::Parameter<Point> * param,
-                                       bool in_global,
-                                       GlobalParamsAction * global_block);
+void Parser::setScalarParameter<RealVectorValue, RealVectorValue>(
```
Comment (permcody) [inline framework/src/parser/Parser.C]: weird, why did clang-format allow this?

**Review by permcody — CHANGES_REQUESTED** [review]
I don't see any major show-stoppers in this PR. I have a few high level comments related to style though. I see several short variable names throughout which I don't like. Even in small scopes, I'd still prefer to see most variables spelled out. Take a look at your dynamic casts + use. I'd suggest that you use more const reference when dealing with strings in general. Block comments? Yeah, I know you can't use your editor to easily comment/uncomment regions of code, but in your documentation strings, you don't typically do that all the time anyway. Consider a header on the contrib files, doesn't necessarily need to be the MOOSE standard header.

**Comment by permcody** [comment]
This PR is not currently Valgrind clean. While I realize that the Parser manages it's own memory, it hands back allocated memory that users are responsible for cleaning. I'm not sure why we'd want to introduce this interface in MOOSE today. Smart pointers and move semantics were designed to eliminate these problems. 

I took the time to use smart pointers and more importantly, move semantics in the code I created in the phase field module, which gets rid of all raw memory **and** smart pointers. Take a look at my data structures here:
https://github.com/idaholab/moose/blob/devel/modules/phase_field/include/postprocessors/FeatureFloodCount.h#L591-L597

The "FeatureData" class that can be stored in a container efficiently without pointers:
https://github.com/idaholab/moose/blob/devel/modules/phase_field/include/postprocessors/FeatureFloodCount.h#L113

I realize that you can't use move semantics in your parse tree because you have several different types stored in your tree, but you should be able to use smart pointers. At the very least, I'd recommend creating an API that hands back smart pointers even if you choose to use raw pointers internally.

File: framework/src/parser/Parser.C (hunk 20 above)
Comment (rwcarlsen) [reply framework/src/parser/Parser.C]: the variable is just a temporary to pass to a c stdlib API.  So using std::array wouldn't really buy anything except making the code have an extra ``.data()`` call when I pass in the var to ``realpath``.  PATH_MAX is a stdlib constant.

File: framework/src/parser/Parser.C (hunk 25 above)
Comment (rwcarlsen) [reply framework/src/parser/Parser.C]: Don't know.

File: framework/src/parser/Parser.C (hunk 20 above)
Comment (permcody) [reply framework/src/parser/Parser.C]: Yes, I realize it's a local. I'm just saying that there are new ways to create stack allocated arrays. It's about style, not function. This is ok here, it's very limited scope.

**Comment by rwcarlsen** [comment]
Alright, comments should be addressed and valgrind should be clean now.

File: test/tests/kernels/forcing_function/forcing_function_error_check.i (hunk 11 above)
Comment (rwcarlsen) [reply test/tests/kernels/forcing_function/forcing_function_error_check.i]: If you look at the test, this is actually suppose to check that the function parser fails when quotes are inside the function text itself.  New parser allows double quotes around strings - so they are no longer treated as literal without surrounding single quotes.

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
Alright, this is mergeable, but I'm going to open up anther ticket for several enhancements I'd like to see in another PR. See #9753.
